### PR TITLE
Impl from string

### PR DESCRIPTION
**Pull request by snewt**
Details on the latest changes to this branch:

Some of the logic implemented in order to validate input and
provide appropriate errors resulted in reverting some of the
parser revisions captured in earlier commits. Lookaheads
are leaned on in order to verify the end of an expression.

If the parser sees `|` that isn't followed by another expression
we can provide an error.

If the parser sees the lhs pattern for a production it can recognize
it rather than attempting to parse it as simply another nonterminal.

Also, the tests that fail on what we're calling "incomplete" input
correspond to a nom error rather than a nom incomplete. If our
parser does its job right the user should never see an incomplete
which would indicate we consumed something we expect later rather than
just the absence of an expected token.

### CODE REVIEW

**Comment by snewt** [comment]
Ah, also added logic to expect eof() in the `from_parse` traits in the form of `*_complete` parsers, i.e. `expression_complete` so that if we don't parse all the input an error can be provided rather than Okay on truncated data.

**Comment by CrockAgile** [comment]
Should this PR be targeting the 0.2.0 branch instead of master?

File: src/parsers.rs [hunk 1]
```diff
@@ -1,70 +1,114 @@
...
     )
 );
 
-
 named!(pub terminal< &[u8], Term >,
     do_parse!(
-        t: delimited!(tag!("\""), take_until!("\""), ws!(tag!("\""))) >>
+        t: alt!(
+            delimited!(char!('"'), take_until!("\""), ws!(char!('"'))) |
+            delimited!(char!('\''), take_until!("'"), ws!(char!('\'')))
```
Comment (CrockAgile) [inline src/parsers.rs]: Nice! This is so readable and handles it well. 👍

**Comment by snewt** [comment]
@CrockAgile yeah I should have been targeting 0.2.0 branch, thanks

**Pull request merged**